In [240]:
# Imports and pip installations (if needed)
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from sklearn import datasets
from sklearn.decomposition import PCA
from sklearn.datasets import load_iris
import pandas as pd
import numpy as np

# import some data to play with
iris = load_iris()

# Part 1: Load the dataset

In [247]:
# Load the dataset (load remotely, not locally)

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.DataFrame(iris.data, columns=iris.feature_names)

In [248]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [250]:
# print the encoding scheme for species: 0 = setosa, 1 = versicolor, 2 = virginica
targets = []
for i in iris.target:
    #print(i)
    if i==0:
        targets.append("setosa")
    elif i==1:
        targets.append("versicolor")
    else:
        targets.append("virginica")
   # targets.append(i)
len(targets)

150

In [252]:
df["Species"] = targets
# Output the first 15 rows of the data
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

   Species  
0   setosa  
1   setosa  
2   setosa  
3   setosa  
4   setosa  
5   setosa  
6   setosa  
7   setosa  
8   setosa  
9   setosa  
10  setosa  
11  setosa  
12  setosa  
13  setosa  
14  setosa

In [187]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [86]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

About the dataset
Explain whats the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?  

Features: the width and length of the sepal, the width and length of the petal
Target: species 
The iris dataset contains the following data
50 samples of 3 different species of iris with 150 samples total
Measurements: sepal length, sepal width, petal length, petal width
the goal is to classify the species of the flower.


# Split the Dataset

In [253]:
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
X= df[['sepal length (cm)','sepal width (cm)','petal length (cm)','petal width (cm)']]
y= df[['Species']] 


In [283]:
# Take the dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.9,test_size=0.1)

# Part 3: Logistic Regression¶


In [284]:
# i. Use sklearn to train a LogisticRegression model on the training set

from sklearn.linear_model import LogisticRegression
clf = LogisticRegression().fit(X_train, y_train)

c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [285]:
clf.predict(X_test)

array(['virginica', 'versicolor', 'versicolor', 'setosa', 'versicolor',
       'virginica', 'versicolor', 'versicolor', 'virginica', 'versicolor',
       'setosa', 'setosa', 'virginica', 'setosa', 'setosa'], dtype=object)

#pick sample datapoints for each sample datapoint

In [286]:
clf.predict_proba(X_test)

array([[4.66296172e-04, 2.98460676e-01, 7.01073027e-01],
       [9.61890510e-03, 8.62861892e-01, 1.27519203e-01],
       [4.24322820e-03, 6.75790936e-01, 3.19965836e-01],
       [9.59634323e-01, 4.03655467e-02, 1.30702338e-07],
       [2.50730601e-03, 8.51039513e-01, 1.46453181e-01],
       [4.51372064e-06, 1.65935640e-02, 9.83401922e-01],
       [2.06838086e-02, 9.25846758e-01, 5.34694337e-02],
       [3.50483071e-03, 8.14148770e-01, 1.82346400e-01],
       [1.23441913e-04, 1.32996018e-01, 8.66880540e-01],
       [3.40592440e-03, 7.30206674e-01, 2.66387402e-01],
       [9.75373002e-01, 2.46269629e-02, 3.53858810e-08],
       [9.51714254e-01, 4.82856219e-02, 1.24366522e-07],
       [5.45013709e-05, 3.77547051e-02, 9.62190794e-01],
       [9.64949605e-01, 3.50503352e-02, 6.00069168e-08],
       [9.73978691e-01, 2.60212776e-02, 3.13688643e-08]])

In [287]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,clf.predict(X_test)))
print("Accuracy:",metrics.accuracy_score(y_test, clf.predict(X_test)))
print("Test Accuracy: " ,clf.score(X_test,y_test))
print("Train Accuracy: " ,clf.score(X_train,y_train))

[[5 0 0]
 [0 6 1]
 [0 0 3]]
Accuracy: 0.9333333333333333
Test Accuracy:  0.9333333333333333
Train Accuracy:  0.9851851851851852


 iii. Report on the score for Logistic regression model, what does the score measure?
The logistic regression is a classification algorithm, so it determines which class or in this case which species the flower belongs too.0.97 represents the accuracy of the algorithm. while .predict is the function that predicts the outcome. In my example, based on the examples I gave it, the logistic regression predicted that the species is setosa

In [288]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
clf.coef_

array([[-0.39465569,  0.99785582, -2.45226597, -1.0447273 ],
       [ 0.57177374, -0.36997092, -0.30165136, -0.76176229],
       [-0.17711805, -0.6278849 ,  2.75391734,  1.80648959]])

In [289]:
clf.intercept_

array([  9.25947063,   2.19472021, -11.45419083])

# Part 4: Support Vector Machine

In [290]:
from sklearn.svm import SVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
#Linear Kernel is used when the data is Linearly separable, that is, it can be separated using a single Line. It is one of the most common kernels to be used. It is mostly used when there are a Large number of Features in a particular Data Set
#Training a SVM with a Linear Kernel is Faster than with any other Kernel
clf_svm = SVC(probability = True)
clf_svm.fit(X_train, y_train)

c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


SVC(probability=True)

In [291]:
# i. Use sklearn to train a Support Vector Classifier on the training set
predictions = clf_svm.predict(X_test)
predictions

array(['virginica', 'versicolor', 'versicolor', 'setosa', 'versicolor',
       'virginica', 'versicolor', 'versicolor', 'virginica', 'versicolor',
       'setosa', 'setosa', 'virginica', 'setosa', 'setosa'], dtype=object)

In [292]:
# ii. For a sample datapoint, predict the probabilities for each possible class
predictions_proba = clf_svm.predict_proba(X_test)

In [293]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,predictions))
print("Accuracy:",metrics.accuracy_score(y_test, predictions))
print("Test Accuracy: " ,clf_svm.score(X_test,y_test))
print("Train Accuracy: " ,clf_svm.score(X_train,y_train))

[[5 0 0]
 [0 6 1]
 [0 0 3]]
Accuracy: 0.9333333333333333
Test Accuracy:  0.9333333333333333
Train Accuracy:  0.9777777777777777


#iii. Report on the score for the SVM, what does the score measure?# iii. Report on the score for the SVM, what does the score measure?
It can solve linear and non-linear problems and work well for many practical problems. The idea of SVM is simple: The algorithm creates a line or a hyperplane which separates the data into classes.So in this case it put the datapoints into one separate class, which is setosa

# Part 5: Neural Network

 iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)


In [320]:
# Experiment with different options for the neural network
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier
solver = ['lbfgs', 'sgd', 'adam']
alpha = [1e-5,1e-4,1e-3]
for i in range(len(solver)):
     clf_mlp = MLPClassifier(solver=solver[i], alpha=alpha[i],hidden_layer_sizes=(4, 2), max_iter = 600,random_state=1)
     clf_mlp.fit(X_train, y_train)
     print("Prediction {} :".format(i),clf_mlp.predict(X_test))
     print("Prediction Score {}:".format(i),clf_mlp.score(X_test, y_test))
     print("==================================================================")

c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Prediction 0 : ['virginica' 'virginica' 'virginica' 'virginica' 'virginica' 'virginica'
 'virginica' 'virginica' 'virginica' 'virginica' 'virginica' 'virginica'
 'virginica' 'virginica' 'virginica']
Prediction Score 0: 0.2
Prediction 1 : ['setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa'
 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa' 'setosa']
Prediction Score 1: 0.3333333333333333


c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Prediction 2 : ['versicolor' 'versicolor' 'versicolor' 'versicolor' 'versicolor'
 'versicolor' 'versicolor' 'versicolor' 'versicolor' 'versicolor'
 'versicolor' 'versicolor' 'versicolor' 'versicolor' 'versicolor']
Prediction Score 2: 0.4666666666666667


c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (600) reached and the optimization hasn't converged yet.
  warnings.warn(


In [321]:
# Experiment with different options for the neural network
clf_mlp.predict([[5.3,0.4,1.1,3]])

array(['versicolor'], dtype='<U10')

In [295]:
clf_mlp.predict_proba(X_test)

array([[0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.84768446, 0.07657605, 0.0757395 ],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.32830457, 0.22391512, 0.44778031],
       [0.84896911, 0.07604211, 0.07498878],
       [0.74177236, 0.11727316, 0.14095447],
       [0.32830457, 0.22391512, 0.44778031],
       [0.84896911, 0.07604211, 0.07498878],
       [0.84896911, 0.07604211, 0.07498878]])

In [296]:
[coef for coef in clf_mlp.coefs_]

[array([[-2.43030234e-07,  1.40831374e-01, -3.29284822e-01,
         -2.82258449e-01, -1.41673050e-01],
        [-2.06886148e-01, -6.14354775e-01, -2.70159875e-03,
         -1.38472156e-01,  1.51784660e-14],
        [-1.05042347e-07,  3.83879304e-01, -8.15574651e-02,
          4.41663793e-01, -2.91904704e-01],
        [ 5.62102807e-03,  1.68686475e-01, -4.46582947e-12,
         -5.40709375e-01, -8.75948757e-02]]),
 array([[ 2.58212688e-01, -2.89066720e-01],
        [-3.61100023e-01, -7.69914404e-01],
        [ 2.33803955e-01, -2.68815873e-01],
        [-1.72557614e-06,  6.78489570e-01],
        [ 1.67428413e-14,  2.33040168e-02]]),
 array([[ 0.26020378, -0.54069681,  0.1221341 ],
        [ 1.68559512, -0.34892266, -1.05745925]])]

In [298]:
clf_mlp.score(X_test, y_test)

0.5333333333333333

In [299]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,clf_mlp.predict(X_test)))
print("Accuracy:",metrics.accuracy_score(y_test,clf_mlp.predict(X_test)))

[[5 0 0]
 [0 0 7]
 [0 0 3]]
Accuracy: 0.5333333333333333


0.5333333333333333

# Part 6: K-Nearest Neighbors

In [305]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=9)
neigh.fit(X_train, y_train)

c:\users\abelf\appdata\local\programs\python\python38\lib\site-packages\sklearn\neighbors\_classification.py:179: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


KNeighborsClassifier(n_neighbors=9)

In [306]:
# iii. Report on the score for the KNN, what does the score measure?
y_train_KNN_predict =neigh.predict(X_train)
y_test_KNN_predict = neigh.predict(X_test)

KNN_train_score = neigh.score(X_train,y_train)
KNN_test_score = neigh.score(X_test,y_test)
print('The accuracy score of the Training model is',KNN_train_score)
print('The accuracy score of the Test model is',KNN_test_score)

The accuracy score of the Training model is 0.9777777777777777
The accuracy score of the Test model is 0.9333333333333333


There is 93% chance that the knn model can assign unknown class point to a certain class.
The Training accuracy rises as the model complexity increases which is determined by the value of K. Larger K value leads to smoother decision boundary (less complex model). Smaller K leads may lead to overfitting). 

Testing accuracy assess if the models overfit or underfit. We get the maximum testing accuracy when the model has right level of complexity, in our case we can see that for a K value of 3 to 19 our model accuracy is 96.6%.

In [302]:
# ii. For a sample datapoint, predict the probabilities for each possible class
neigh.predict_proba(X_test)

array([[0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [0., 0., 1.],
       [1., 0., 0.],
       [1., 0., 0.]])

In [304]:
print(classification_report(y_train,y_train_KNN_predict))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        45
  versicolor       0.95      0.95      0.95        43
   virginica       0.96      0.96      0.96        47

    accuracy                           0.97       135
   macro avg       0.97      0.97      0.97       135
weighted avg       0.97      0.97      0.97       135



 # Part 7: Conclusions and takeaways
In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

In [309]:
print("logistic regression score", clf.score(X_test,y_test))
print("Support Vector Machine score", clf_svm.score(X_test,y_test))
print("Neural Network score", clf_mlp.score(X_test,y_test))
print("KNN score", KNN_test_score)

logistic regression score 0.9333333333333333
Support Vector Machine score 0.9333333333333333
Neural Network score 0.5333333333333333
KNN score 0.9333333333333333


Logistic Regression, SVM, and KNN performed equally while mlp classfier performed the lowest. A neural network model for binary classification can be worse than a logistic regression model because neural networks are more difficult to train and are more prone to overfitting than logistic regression, which explains why we are getting a lower score for neural net.